### DISCUSSION TITLE
Throwing Custom exceptions

### DISCUSSION THREAD

**Discussion by lars-lindstrom** [ardalis/GuardClauses#181] [Q&A]
I have previously used the Ardalis Specification/Repository assembly with great pleasure and currently the GuardClause assembly got my attention. 
 In general it is great and I love it. But listening to your podcasts, inspired me to use custom exceptions to a greater extend. Unfortunately I have fails to identify a way to either use functionality within Ardalis GuardClauses or to make and overloaded version myself.
What I’m looking for I something like this: Guard.Agains.Null<CustomException>(null object)

**Comment by ardalis** [comment]
I would use the extension approach shown in the project README. Don't just call it `Guard.Against.Null()` though but something more specific like `Guard.Against.MissingOrder(Order order)` so that it's more specific and obvious when the problem occurs (that's the point of my prefer custom exceptions advice).

https://ardalis.com/prefer-custom-exceptions-to-framework-exceptions/

## Extending with your own Guard Clauses
To extend your own guards, you can do the following:

```csharp
// Using the same namespace will make sure your code picks up your 
// extensions no matter where they are in your codebase.
namespace Ardalis.GuardClauses
{
    public static class FooGuard
    {
        public static void Foo(this IGuardClause guardClause, string input, string parameterName)
        {
            if (input?.ToLower() == "foo")
                throw new ArgumentException("Should not have been foo!", parameterName);
        }
    }
}

// Usage
public void SomeMethod(string something)
{
    Guard.Against.Foo(something, nameof(something));
}
```
